Assets/MRTK/StandardAssets/EditorUtilities/OnLoadUtilities.cs

```diff
@@ -0,0 +1,113 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System.Collections.Generic;
+using System.IO;
+using UnityEditor;
+using UnityEngine;
+
+namespace Microsoft.MixedReality.Toolkit.Utilities.Editor
+{
+    [InitializeOnLoad]
+    public static class OnLoadUtilities
+    {
+        private const string SessionStateKey = "StandardAssetsOnLoadUtilitiesSessionStateKey";
+        
+        private const string ShaderSentinelFile = "MRTK.Shaders.Sentinel";
+        private const string ShaderImportDestination = "MRTK/Shaders";
+
+        static OnLoadUtilities()
+        {
+            // This InitializeOnLoad handler only runs once at editor launch in order to adjust for Unity version
+            // differences. These don't need to (and should not be) run on an ongoing basis. This uses the
+            // volatile SessionState which is clear when Unity launches to ensure that this only runs the
+            // expensive work (file system i/o) once.
+            if (!SessionState.GetBool(SessionStateKey, false))
+            {
+                SessionState.SetBool(SessionStateKey, true);
+                EnsureShaders();
+            }
+        }
+
+        /// <summary>
+        /// Ensures that MRTK shader files are present in a writable location. To support the 
+        /// Universal Render Pipeline, shader modifications must be persisted.
+        /// </summary>
+        private static void EnsureShaders()
+        {
+            if (!AssetsContainsShaders())
+            {
+                ImportShaderFiles();
+            }
+        }
+
+        /// <summary>
+        /// Checks to see if the Assets or Packages (if embedded) folder trees contains the MRTK shaders.
+        /// </summary>
+        /// <returns>True if the shader sentinel file is found, otherwise false.</returns>
+        private static bool AssetsContainsShaders()
+        {
+            List<string> searchFolders = new List<string>
```

Comment:
We chatted a bit offline about this, but I wanted to publicly document it as well.

This method assumes that embedded packages will be from the .unitypackage distribution and not the npm tgz distribution. Since the tgz will have the shader in a `Shaders~` folder, this search will succeed (the sentinel exists in the file structure) but the shader isn't actually visible to Unity. This situation is probably a bit of an edge case, but I think using [`StandardShaderUtility.MrtkStandardShader`](https://github.com/microsoft/MixedRealityToolkit-Unity/blob/94fffc24775ee8d9f5a500192eb33a02c2663283/Assets/MRTK/Core/Utilities/StandardShader/StandardShaderUtility.cs#L21) to check if Unity can see the shader would account for this (and all other?) cases.

Reply:
I'll look into that in a bit more detail,.

Reply:
need to confirm the location of the function and whether or not it is safe to move to a new assembly

Reply:
Good call, I didn't think about it being in a different package.

Could just use the underlying `Shader.Find` in that case. The more Unity API calls vs searching the file structure manually the better, imo!

Reply:
i currently have an implementation using AssetDatabase.GUIDToAssetPath(sentinel) do you have a preference?